lib/network.py

```diff
@@ -1422,19 +1412,32 @@ def try_to_recover(err):
             self.print_error("wait_on_sockets: {} raised by select() call.. trying to recover...".format(err))
             self.find_bad_fds_and_kill()
 
+        rin = []
+        win = []
+        r_immed = []
         with self.interface_lock:
             interfaces = list(self.interfaces.values())
-            rin = [i for i in interfaces if i.fileno() > -1]
-            win = [i for i in interfaces if i.num_requests() and i.fileno() > -1]
+            for interface in interfaces:
+                if interface.fileno() < 0:
+                    continue
+                read_pending, write_pending = interface.pipe.get_selectloop_info()
+                if read_pending:
+                    r_immed.append(interface)
+                else:
+                    rin.append(interface)
+                if write_pending or interface.num_requests():
+                    win.append(interface)
 
-        # Python docs say Windows doesn't like empty selects.
-        # Sleep to prevent busy looping
-        if not win and not rin:
-            time.sleep(0.1)
-            return
+        timeout = 0 if r_immed else 0.1
 
         try:
-            rout, wout, xout = select.select(rin, win, [], 0.1)
+            # Python docs say Windows doesn't like empty selects.
+            if win or rin:
+                rout, wout, xout = select.select(rin, win, [], timeout)
+            else:
+                # Sleep to prevent busy looping
+                time.sleep(timeout)
```

Comment:
I wonder if this sleep here should be condifional on `not r_immed`?  This way if there are `r_immed` no sleep occurs?  Thoughts?

Reply:
Sure that's possible :+1: